### PR TITLE
support basic toc parser and build

### PR DESCRIPTION
**Pull request by OsmondJiang**
1. md and yaml toc parser
2. toc build
3. resolve basic href from current docset only
4. unit test and e2e test

### CODE REVIEW

File: docs/specs/build/toc/basics.yml [hunk 1]
```diff
@@ -0,0 +1,87 @@
+---
```
Comment (yufeih) [inline docs/specs/build/toc/basics.yml]: split the test into multiple small tests

File: src/docfx/build/Build.cs [hunk 2]
```diff
@@ -30,12 +32,31 @@ private static List<Document> GlobFiles(Context context, Docset docset)
                            .ToList();
         }
 
-        private static Task BuildFiles(Context context, List<Document> files, TableOfContentsMap tocMap)
+        private static Task BuildFiles(Context context, List<Document> files, TableOfContentsMap tocMap, bool watch = false)
```
Comment (yufeih) [inline src/docfx/build/Build.cs]: remove `watch`

File: src/docfx/build/toc/BuildTableOfContents.cs [hunk 3]
```diff
@@ -11,9 +11,26 @@ namespace Microsoft.Docs.Build
 {
     internal static class BuildTableOfContents
     {
-        public static Task Build(Context context, Document file)
+        public static Task Build(Context context, Document file, Action<Document> buildChild)
         {
-            throw new NotImplementedException();
+            if (file.ContentType != ContentType.TableOfContents)
```
Comment (yufeih) [inline src/docfx/build/toc/BuildTableOfContents.cs]: Debug.Assert(file.ContentType == TOC)

File: src/docfx/build/toc/BuildTableOfContents.cs [hunk 4]
```diff
@@ -11,9 +11,26 @@ namespace Microsoft.Docs.Build
...
+            }
+
+            var (tocModel, refArticles, refTocs) = Load(file);
+
+            foreach (var article in refArticles)
+            {
+                buildChild(article);
+            }
+
+            if (tocModel != null)
```
Comment (yufeih) [inline src/docfx/build/toc/BuildTableOfContents.cs]: when will tocModel be null?

File: src/docfx/build/toc/BuildTableOfContents.cs [hunk 5]
```diff
@@ -44,9 +61,44 @@ private static Task BuildTocMap(Context context, Document fileToBuild, TableOfCo
             return Task.CompletedTask;
         }
 
-        private static (TableOfContentsModel tocModel, List<Document> referencedDocument, List<Document> referencedTocs) Load(Document fileToBuild)
+        private static (List<TableOfContentsModel> tocModel, List<Document> referencedDocument, List<Document> referencedTocs) Load(Document fileToBuild)
         {
-            throw new NotImplementedException();
+            if (fileToBuild == null)
```
Comment (yufeih) [inline src/docfx/build/toc/BuildTableOfContents.cs]: this should never be null

File: test/docfx.Test/common/TestHelper.cs [hunk 6]
```diff
@@ -9,56 +9,56 @@ namespace Microsoft.Docs.Build
 {
     internal static class TestHelper
     {
-        public static TheoryData<string, string> FindTestSpecs(string path)
+        public static TheoryData<TestSpec> FindTestSpecs(string path)
```
Comment (yufeih) [inline test/docfx.Test/common/TestHelper.cs]: Please help revert these changes, I intentionally use `TheoryData<string, string>` to avoid YAML deserialization on xunit test discovery

File: src/docfx/build/toc/BuildTableOfContents.cs (hunk 5 above)
Comment (OsmondJiang) [reply src/docfx/build/toc/BuildTableOfContents.cs]: removed

---
In reply to: [185712505](https://github.com/dotnet/docfx/pull/2717#discussion_r185712505) [](ancestors = 185712505)

File: docs/specs/build/toc/basics.yml [hunk 7]
```diff
@@ -0,0 +1,596 @@
...
+                      "href":"n1"
+                  }
+                ],
+                "toc_title":"Existing File Reference"
+            }
+          ],
+          "toc_title":"Index Reference",
+          "href":"."
+      }
+    ]
```
Comment (superyyrrzz) [inline docs/specs/build/toc/basics.yml]: toc.json is an object instead of an array

File: stylecop.ruleset [hunk 8]
```diff
@@ -21,6 +21,8 @@
     <Rule Id="SA1311" Action="None" />
     <!-- SA1401FieldsMustBePrivate -->
     <Rule Id="SA1401" Action="None" />
+    <!-- SA1402FileMayOnlyContainASingleType -->
+    <Rule Id="SA1402" Action="None" />
```
Comment (superyyrrzz) [inline stylecop.ruleset]: Do not enable this for easy locating source of a type

File: src/docfx/build/toc/BuildTableOfContents.cs [hunk 9]
```diff
@@ -44,9 +55,39 @@ private static Task BuildTocMap(Context context, Document fileToBuild, TableOfCo
             return Task.CompletedTask;
         }
 
-        private static (TableOfContentsModel tocModel, List<Document> referencedDocument, List<Document> referencedTocs) Load(Document fileToBuild)
+        private static (List<TableOfContentsModel> tocModel, List<Document> referencedDocument, List<Document> referencedTocs) Load(Document fileToBuild)
```
Comment (superyyrrzz) [inline src/docfx/build/toc/BuildTableOfContents.cs]: >referencedDocument [](start = 76, length = 18)

referencedDocuments

File: stylecop.ruleset (hunk 8 above)
Comment (OsmondJiang) [reply stylecop.ruleset]: you mean do not disable?

---
In reply to: [185723267](https://github.com/dotnet/docfx/pull/2717#discussion_r185723267) [](ancestors = 185723267)

File: docs/specs/build/toc/basics.yml (hunk 7 above)
Comment (OsmondJiang) [reply docs/specs/build/toc/basics.yml]: ho god, thanks for the reminding!

---
In reply to: [185722737](https://github.com/dotnet/docfx/pull/2717#discussion_r185722737) [](ancestors = 185722737)

File: stylecop.ruleset (hunk 8 above)
Comment (superyyrrzz) [reply stylecop.ruleset]: yes

---
In reply to: [185723428](https://github.com/dotnet/docfx/pull/2717#discussion_r185723428) [](ancestors = 185723428,185723267)

Comment (OsmondJiang) [reply stylecop.ruleset]: enabled and split states and rules

---
In reply to: [185723572](https://github.com/dotnet/docfx/pull/2717#discussion_r185723572) [](ancestors = 185723572,185723428,185723267)
